### ISSUE TITLE
Warnings or errors when printing

### ISSUE THREAD

**Issue by jad-hamza** [epfl-lara/stainless#103]
These are various warnings or errors I get when using println.

```scala
object SI_Print1 {
  println("test")
}
```
> [Warning] Could not extract tree in static container: scala.this.Predef.println("test")

```scala
object SI_Print2 {
  def test() = {
    println("test")
  }
}
```
> [Internal] Missing some nodes in Registry: println$4

```scala
object SI_Print3 {
  stainless.io.StdOut.print(5)
}
```
> [Warning] Could not extract tree in static container: stainless.io.StdOut.print(5)

`SI_Print3` actually works fine when the print is not at the top-level.

```scala
import stainless.annotation._

object SI_Print4 {
  @library
  def pp(a: Any) = println(a)
}
```
> [Internal] Missing some nodes in Registry: println$4

```scala
import stainless.annotation._

object SI_Print5 {
  @library
  @extern
  def pp(a: Any) = println(a)
  pp("Hello")
}
```
> [Warning] Could not extract tree in static container: SI_Print4.this.pp("Hello")

 ```scala
import stainless.annotation._

object SI_Print6 {
  @library
  @extern
  def pp(a: Any) = println(a)

  def test() = {
    pp("Hello")
  }
}
```

> pp("Hello") is of type <untyped>
>  "Hello" is of type String because pp was instantiated with  with type (Object) => Unit

**Comment by jad-hamza** [comment]
Just noticed a typo in the last error message: "with with", from Inox:
https://github.com/epfl-lara/inox/blob/master/src/main/scala/inox/ast/SymbolOps.scala#L1265-L1267

**Comment by samarion** [comment]
Well, I can't really call these _bugs_...

1. You can't define any state mutation within an object or class body. The imperative transformations don't support this. You could theoretically call a pure function there, but that doesn't really make much sense to we just don't allow any calls within a body's definition. I guess the warning could be made a bit clearer.
2. `println` is not defined within the Stainless library and therefore can't be extracted unless we add dedicated support.
3. Same problem as 1.
4. Same problem as 2.
5. Same problem as 1.
6. This is indeed a bug, but bugs in Scala's type system support are expected since I've only had time to implement about half of it.

It could be useful to add a warning if a user tries to use `println` explaining that/why it is not supported in Stainless. This should be a fairly simple extension to the frontends (see [here](https://github.com/epfl-lara/stainless/blob/3e8aa4485761908531f49cab606806bc21f4a777/frontends/scalac/src/main/scala/stainless/frontends/scalac/CodeExtraction.scala#L1263) for an example in that direction, though the println warning should take place in `extractTree`).

**Comment by jad-hamza** [comment]
Bugs in the reporting :)

Can print be assumed pure from a verification point of view?

**Comment by jad-hamza** [comment]
This is perhaps related, this program creates a NullPointerException:
```scala
object SI_Print7 {
  5
}
```

**Comment by mantognini** [comment]
The last example is addressed by #159.

**Comment by vkuncak** [comment]
Hi Jad,
What is the use case of println for you? You want to debug code using stainless evaluator, scala repl, or something else? I would prefer to use toString methods in most cases.

**Comment by jad-hamza** [comment]
I think I encountered these errors and warnings when writing https://github.com/epfl-lara/bolts/blob/55b5bfd3ab1e656cce1a6b686487e968bb1f717f/software-foundations/sf1-logical-foundations/Basics.scala

(They get displayed when compiling with scalac and running with scala as described in the README: https://github.com/epfl-lara/bolts/tree/master/software-foundations)

But otherwise I don't need this feature or rely on it. And the evaluator of parameterless functions covers the debugging usecase. We could add rules that explicitly reject the print/println from Scala

**Comment by romac** [comment]
> We could add rules that explicitly reject the print/println from Scala

This has been fixed in 4c61eddbbf74bc5207f8f712dde14c694f7d36ae.
